servers/src/main/java/tachyon/worker/block/evictor/LFUEvictor.java

```diff
@@ -0,0 +1,206 @@
+   * Node, "-" represents bi-directional link
+   *
+   * 1 - o - o
+   * |
+   * 2 - o
+   * |
+   * ...
+   *
+   * When a Node is accessed, move it to next layer, if next layer represents the same access count
+   * with this node, append it to the end of next layer, otherwise, create a new next layer with the
+   * Node as the only element
+   *
+   * If the access count arrives at limit of {@code long}, it will no longer be increased
+   */
+  class CountNode extends DoubleLinkListNode {
+    long count;
+    Node first, last;
+
+    CountNode() {
+      count = 0;
+      prev = next = null;
+      first = last = null;
+    }
+
+    CountNode(long count) {
+      this.count = count;
+      prev = next = null;
+      first = last = null;
+    }
+
+    void appendNode(Node node) {
+      if (first == null) {
+        first = node;
+        node.prev = first;
+      } else {
+        last.append(node);
+        last = node;
+      }
+      last = node;
+    }
+
+    CountNode nextCountNode() {
+      return (CountNode)next;
+    }
+  }
+  class Node extends DoubleLinkListNode {
+    long blockId;
+    CountNode head;
+
+    Node() {
+      blockId = -1;
+      prev = next = null;
+      head = null;
+    }
+
+    Node(long id) {
+      blockId = id;
+      prev = next = null;
+      head = null;
+    }
+
+    // assume next layer exists and have elements other than the count node
+    void appendToNextLayer() {
+      remove();
+      nextLayerHead().appendNode(this);
+    }
+
+    long count() {
+      return head.count;
+    }
+
+    Node first() {
+      return head.first;
+    }
+
+    CountNode nextLayerHead() {
+      return head.nextCountNode();
+    }
+
+    Node nextNode() {
+      return (Node)next;
+    }
+  }
+
+  /** Always be the head of the first layer */
+  private CountNode mHead;
+  /** Map from blockId to corresponding Node */
+  private Map<Long, Node> mCache;
+
+  private BlockMetadataManager mMeta;
+
+  public LFUEvictor(BlockMetadataManager meta) {
+    mMeta = meta;
+
+    mHead = new CountNode();
+    mCache = new HashMap<Long, Node>();
+  }
+
+  /** Complexity: O(1) */
+  @Override
+  public void onAccessBlock(long userId, long blockId) {
+    synchronized (mMeta) {
+      if (mCache.containsKey(blockId)) {
+        Node node = mCache.get(blockId);
+        if (node.count() == Long.MAX_VALUE) {
+          return;
+        }
+        if (node.nextLayerHead() == null && node.first() == node) { // only Node in last layer
+          node.head.count += 1;
+        } else {
+          if (node.nextLayerHead() == null || node.nextLayerHead().count != node.count() + 1) {
+            // create new layer
+            node.head.append(new CountNode(node.count() + 1));
+          }
+          node.appendToNextLayer();
+        }
+      } else {
+        mHead.appendNode(new Node(blockId));
+      }
+    }
+  }
+
+  /**
+   * Evict layer by layer, in each layer, evict from first to last
+   *
+   * Complexity: O(1)
```

Comment:
I mean the complexity of each time choosing a least frequently accessed candidate  is O(1).  should remove it.